Move reusable functions to helpers folder

aminya opened the pull request:
Moved reusable functions to src folder. It makes expanding and developing the package easier. 

This may be temporary until we reach a stable API.

tshort commented:
I'm not a fan of moving temporary stuff into the core code. We'd never want the code you moved in the base part of the package. Eventually, we'll want some compiler helpers, but it won't be this code. 

If you're looking to make it easier to test out on different code, I'd prefer a script under `test` or `examples` that pulls in code needed.

aminya commented:
> 
> 
> I'm not a fan of moving temporary stuff into the core code. We'd never want the code you moved in the base part of the package. Eventually, we'll want some compiler helpers, but it won't be this code.
> 
> If you're looking to make it easier to test out on different code, I'd prefer a script under `test` or `examples` that pulls in code needed.

I can make a separate module (or package) if you want. 

The reason I find moving the code to src or some package useful is that we clearly see which functions are reusable and so we can work on them (or even use them in other works).

This PR is not just code-moving, I also created some functions that do specific tasks.

tshort commented:
I prefer one of the following:

* separate package
* `helpers` directory somewhere not under `/src`--The readme can include directions on how to load the helpers.

aminya commented:
> 
> 
> I prefer one of the following:
> 
>     * separate package
> 
>     * `helpers` directory somewhere not under `/src`--The readme can include directions on how to load the helpers.

I will make a separate module under `root/helpers` directory.

aminya commented:
Done!

tshort commented on src/StaticCompiler.jl:
```diff
@@ -21,4 +21,8 @@ include("globals.jl")
 include("overdub.jl")
 include("irgen.jl")
 
+include("../helpers/jlrun.jl")
```
I'd rather not include the code into the core part of the package. Because doing that pollutes the `Project.toml` (with `Test`) and exports stuff I really don't want exported.

I'd rather you provide a script in `/helpers` that defines the code you want, and the user can do `include("something/StaticCompiler/helpers/load_helpers.jl")`.

aminya replied:
Loading files manually isn't the best way to use a code. Any external user that installs a package should be able to call the functions easily when they write `using StaticCompiler`.

In this case, I think I can move the code to a separate package called `CompilerUtils`.

aminya replied:
It is not possible or (is a bad practice) to have two packages that use each other's functions.
We will get this error:
https://julialang.slack.com/archives/C67EFTEF3/p1576787124006700

As discussed here, @KristofferC suggest just to keep the code in the same package.
https://julialang.slack.com/archives/C67EFTEF3/p1576789543011200

tshort replied:
It can be a separate package if CompilerUtils depends on StaticCompiler. With that, the code in StaticCompiler under `/test` could stay as is. You could repeat that code in CompilerUtils and use it as a testbed for better APIs for compiling.

If you want to keep everything under StaticCompiler under `/helpers`, that's fine, too. I just don't want stuff pulled into the main code until it has a better API.

aminya replied:
I achieved the best of both worlds using git submodules. Now, this can be merged!

tshort replied:
I don't understand how this works. You might have to explain...

aminya replied:
Basically, we can link two repositories together using git submodules. This is one way to have multiple parts of a big project in different repositories. 

Now, `helpers` folder contains CompilerUtils repository, and so when we clone StaticCompiler, this folder that includes CompilerUtils is cloned too. 

You can learn more here:
https://gist.github.com/gitaarik/8735255

KristofferC replied:
Why not put the code directly into `helpers`? Submodules have many drawbacks, for example, you don't get the files of a submodule when you clone, you need to explicitly get them.

aminya replied:
@KristofferC the code is only needed for `runtests`, and CI automatically clones the submodules. So there is no issue in this case. 

For local test, one can run this to initialize the submodule:
```
git submodule update --init --recursive
```

To update the submodule if CompilerUtils changes, one should run:
```
git submodule update --remote --merge
```

tshort replied:
Based on Kristoffer's comments, I'd prefer the code go back in `helpers`. I'm also willing to  (temporarily) `include` that into the code under `src` if you can remove the dependency on `Test`.

aminya replied:
That is what I did in the first place. I will remove CompilerUtils and add back the code here.

tshort replied:
I realize that. It's become the least bad option. Sorry for the runaround...

tshort commented on Project.toml:
```diff
@@ -15,7 +14,8 @@ TypedCodeUtils = "687fb87b-adea-59d5-9be9-82253b54685d"
 
 [targets]
-test = ["Test"]
+test = ["Test", "Formatting"]
```
Good idea here.

aminya commented:
Test passes locally now. I don't understand the problem with `@jlrun` on Travis
https://travis-ci.com/aminya/StaticCompiler.jl/jobs/269406893#L223

tshort commented:
On Linux, tests fail for me locally, too.

tshort commented:
It's a weird macro error. `@jlrun` works in the REPL, but inside a `@testset`, it fails.

tshort commented:
@aminya, any idea what changed since the last time you tried this approach?

aminya commented:
> It's a weird macro error. `@jlrun` works in the REPL, but inside a `@testset`, it fails.

That is very strange. It might be related to path of the files because @__DIR__ is used in the `jlrun` macro.

aminya commented:
I thought it might be because of the old gcc compiler on Linux vs the new one on Windows. I added a Linux Bionic test. However, it is not the case

Also, I added MacOS. It fails on the same code, but with a slightly different error:
Linux - 3 of them fails
https://travis-ci.com/aminya/StaticCompiler.jl/jobs/269898292#L222
MacOS - 1 of them fails
https://travis-ci.com/aminya/StaticCompiler.jl/jobs/269898293#L188

tshort commented:
The failing tests are bizarre, especially since you had a previous version that I thought worked.

It's also a sign that `@jlrun` isn't very robust.

aminya commented:
I have created issues for the problems. I think this should be merged for now and fix the issues in other PRs.
https://github.com/tshort/StaticCompiler.jl/issues/12
https://github.com/tshort/StaticCompiler.jl/issues/13

tshort commented:
I'd rather not revert tests. It'd be better to fix things first.

aminya commented:
@tshort so should we merge this the same way you decided in https://github.com/tshort/StaticCompiler.jl/pull/14#issuecomment-570783058

tshort commented:
The difference is that these test failures are reproducible. I'll try and see if I can identify the root problem.

tshort commented:
When run with `@jlrun`, the libc error code is changed to 4 (interrupted system call). If it's flipped around, the test passes. This works: `@test (@jlrun f1()) == f1()`.

Based on that, I'm inclined to merge, but I wanr to poke around more.

tshort commented:
Let's work on fixes on master. The top priority is fixing Mac tests or removing them.

aminya commented:
> 
> 
> When run with `@jlrun`, the libc error code is changed to 4 (interrupted system call). If it's flipped around, the test passes. This works: `@test (@jlrun f1()) == f1()`.
> 
> Based on that, I'm inclined to merge, but I want to poke around more. It might be better to remove those rather than use `test_skip`.
I'm glad you found the source of the error!

tshort commented:
Well, crossing fingers didn't help. The issue from #14 is popping up on master after this merge. Tests are only intermittently passing.
